[All] Plugins finding and loading

guparan opened the pull request:
This PR aims to improve the way we search and load plugins.

It adds:
- recursive search in PluginManager::findPlugin
- finding plugin dependencies in its directory (Windows only)
- (SofaPython) function to easily edit sys.path at any plugin init
- ~~new Data in RequiredPlugin to find out-of-sofa plugins: `searchPath`~~
- new component AddPluginRepository for adding a path to the pool of plugin paths
- ~~**BREAKING**~~: rename AddResourceRepository into AddDataRepository (alias added for backward compat)
- new default plugins folder in build/install to allow drag&drop

The last one shall be discussed, I'm not sure it is really needed.

______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

guparan committed:
[SofaPython] ADD module loading from plugin name

guparan committed:
[SofaPython] Improve module loading from plugin name

guparan committed:
[SofaHelper] FIX dll subdependency loading

LOAD_WITH_ALTERED_SEARCH_PATH was needed for runSofa to find the dependencies of out-of-tree installed plugins

guparan committed:
[SofaHelper] ADD recursive search in PluginManager::findPlugin

guparan committed:
[SofaHelper] ADD backslashes replacement in FileRepository::cleanPath

guparan committed:
[SofaGraphComponent] ADD searchPath Data in RequiredPlugin

User-given directory to scan before the default ones when searching for the plugin.

# Conflicts:
#	modules/SofaGraphComponent/RequiredPlugin.cpp

guparan committed:
[CMake] ADD a default plugins directory in build/install

# Conflicts:
#	applications/projects/runSofa/Main.cpp

damienmarchal commented on modules/SofaGraphComponent/RequiredPlugin.cpp:
```diff
@@ -41,19 +43,32 @@ int RequiredPluginClass = core::RegisterObject("Load the required plugins")
     , d_requireAll ( initData(&d_requireAll , true, "requireAll", "Display an error message if any plugin names failed to be loaded"))
 {
     this->f_printLog.setValue(true); // print log by default, to identify which pluging is responsible in case of a crash during loading
+    d_pluginName.setRequired(true);
```
The name is not required, the altenrative is to set he pluginName in the name field which allows to write:
<RequiredPlugin name="Toto"/> instead of <RequiredPlugin pluginName="Toto"/>
The nice thing by using <RequiredPlugin name="Toto"/> is that it display in the scene graph view of runSofa the name of the plugin.

EDIT: I see your added warning after

guparan replied:
I reverted this change as I understood the interest of having the same value for both `name` and `pluginName`: console output and GUI are more user-friendly this way.

epernod commented on SofaKernel/framework/sofa/helper/system/DynamicLibrary.cpp:
```diff
@@ -63,7 +64,9 @@ const std::string& DynamicLibrary::Handle::filename() const
 # if defined(WIN32)
-    void *handle = ::LoadLibraryA(filename.c_str());
+    std::string p(filename);
+    std::replace(p.begin(), p.end(), '/', '\\'); // ensure Windows style path
```
is this still really needed in 2019 ?

epernod replied:
I see you define this old trick as well in cleanPath. Maybe will be more consistent to call 
`LoadLibraryExA(cleanPath(p).c_str(), ...`

guparan replied:
I followed your advice but also deprecated FileRepository::cleanPath as it basically was a duplicate of FileSystem::cleanPath.

epernod commented on SofaKernel/framework/sofa/helper/system/PluginManager.h:
```diff
@@ -168,7 +168,7 @@ class SOFA_HELPER_API PluginManager
     void init(const std::string& pluginPath);
 
-    std::string findPlugin(const std::string& pluginName, const std::string& suffix = getDefaultSuffix(), bool ignoreCase = true);
+    std::string findPlugin(const std::string& pluginName, const std::string& suffix = getDefaultSuffix(), bool ignoreCase = true, int maxRecursiveDepth = 4);
```
magic number or is 4 really meaning something?

guparan replied:
Changed it to 6 :wink:

damienmarchal commented:
Hi Guillaume,

Thanks for the PR, contains useful things. About the "plugins" directory I personally like the idea of stopping to mix shared libraries with sofa plugin, so it is all right to me.

guparan committed:
[SofaHelper] FIX compatibility with boost < 1.60

guparan committed:
Revert "[SofaGraphComponent] ADD searchPath Data in RequiredPlugin"

This reverts commit e93d063ea392a2d9f994d7bc8cc0440be4582f9e.

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into plugins_finding_and_loading

guparan committed:
[SofaHelper] CLEAN DynamicLibrary nullptr usage

guparan committed:
[SofaGraphComponent] Use case-insensitive recursive search for plugin loading

guparan committed:
[SofaMisc] Rework AddResourceRepository API

- AddResourceRepository changed to an abstract class
- AddDataRepository component added (same behavior as old AddResourceRepository)
- AddPluginRepository component added

guparan committed:
[SofaHelper] Deprecate FileRepository::cleanPath

Use FileSystem::cleanPath instead

guparan committed:
[SofaHelper] ADD recursive parameter to PluginManager::loadPlugin

guparan committed:
[SofaMisc] FIX AddResourceRepository_test

guparan committed:
[SofaHelper] FIX PluginManager on CentOS

guparan commented:
[ci-build][with-all-tests]

guparan committed:
[SofaGraphComponent] Allow paths in RequiredPlugin

guparan committed:
[SofaHelper] FIX FileSystem::cleanPath replace size

guparan committed:
[SofaMisc] CLEAN AddResourceRepository

guparan committed:
[SofaMisc] RENAME AddResourceRepository to BaseAddResourceRepository

damienmarchal commented on SofaKernel/framework/sofa/helper/system/FileRepository.cpp:
```diff
@@ -127,26 +127,16 @@ FileRepository::~FileRepository()
-        pos = p.find("//");
-    }
-    return p;
+    msg_deprecated("FileRepository::cleanPath") << "Use FileSystem::cleanPath instead.";
```
I think you should use dmsg_deprecated instead of msg_deprecated
and also use 
with https://en.cppreference.com/w/cpp/language/attributes/deprecated

damienmarchal commented:
Read ? Go.
